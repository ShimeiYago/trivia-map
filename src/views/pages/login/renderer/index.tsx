import { ADMIN_LINK } from 'constant/links';
import React from 'react';
import { Navigate } from 'react-router-dom';
import { sleep } from 'utils/sleep';
import { AuthForms } from 'views/components/organisms/auth-forms';

export class Renderer extends React.Component<unknown, State> {
  state: State = {
    redirect: false,
  };

  render() {
    if (this.state.redirect) {
      return <Navigate to={ADMIN_LINK} />;
    }

    return (
      <AuthForms
        initialMode="login"
        onLoginSucceed={this.redirectAfterSeveralSeconds}
      />
    );
  }

  protected redirectAfterSeveralSeconds = async () => {
    await sleep(1000);
    this.setState({
      redirect: true,
    });
  };
}

export type State = {
  redirect: boolean;
};
