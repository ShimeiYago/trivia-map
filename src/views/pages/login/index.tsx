import { Renderer } from './renderer';

export const Login = Renderer;
