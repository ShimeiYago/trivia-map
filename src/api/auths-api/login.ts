import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { mockLoginResponse } from 'api/mock/auths-response/login';
import { User } from 'types/user';

export async function login(
  email: string,
  password: string,
): Promise<LoginResponse> {
  const axiosInstance = getAxiosInstance({}, mockLoginResponse);

  const requestData: LoginRequest = {
    email: email,
    password: password,
  };

  try {
    // TODO: Set reasonable timeout
    const res: AxiosResponse<LoginResponse> = await axiosInstance.post(
      `${BASE_URL}/auths/login`,
      requestData,
    );
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type LoginRequest = {
  email: string;
  password: string;
};

export type LoginResponse = {
  access_token: string;
  refresh_token: string;
  user: User;
};

export type ValidationError = {
  email?: string[];
  password?: string[];
  non_field_errors?: string[];
};
