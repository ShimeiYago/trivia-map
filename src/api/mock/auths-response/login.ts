import { LoginResponse } from 'api/auths-api/login';

export const mockLoginResponse: LoginResponse = {
  access_token: 'xxx',
  refresh_token: 'xxx',
  user: {
    userId: 1,
    email: 'user@example.com',
    nickname: 'Axel',
    icon: 'https://disneyparkstory.com/wp-content/uploads/2017/09/apple-touch-icon.png',
  },
  access_token_expiration: '2023-10-05T01:31:10.861050+09:00',
  refresh_token_expiration: '2023-10-05T01:31:10.861050+09:00',
};
