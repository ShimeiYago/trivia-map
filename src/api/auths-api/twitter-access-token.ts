import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { mockTwitterAccessTokenResponse } from 'api/mock/auths-response/twitter-access-token';

export async function twitterAccessToken(
  requestData: TwitterAccessTokenRequest,
): Promise<TwitterAccessTokenResponse> {
  const axiosInstance = getAxiosInstance({}, mockTwitterAccessTokenResponse);

  try {
    const res: AxiosResponse<TwitterAccessTokenResponse> = await axiosInstance.post(
      `${BASE_URL}/auths/twitter/access-token`,
      requestData,
    );
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}

export type TwitterAccessTokenRequest = {
  oauthToken: string;
  oauthVerifier: string;
};

export type TwitterAccessTokenResponse = {
  accessToken: string;
  accessTokeSecret: string;
};
