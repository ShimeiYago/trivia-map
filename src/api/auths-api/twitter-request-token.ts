import { mockTwitterRequestTokenResponse } from './../mock/auths-response/twitter-request-token';
import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function twitterRequestToken(
  requestData: TwitterRequestTokenRequest,
): Promise<TwitterRequestTokenResponse> {
  const axiosInstance = getAxiosInstance({}, mockTwitterRequestTokenResponse);

  try {
    const res: AxiosResponse<TwitterRequestTokenResponse> = await axiosInstance.post(
      `${BASE_URL}/auths/twitter/request-token`,
      requestData,
    );
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}

export type TwitterRequestTokenRequest = {
  callbackUrl: string;
};

export type TwitterRequestTokenResponse = {
  authenticateUrl: string;
};
