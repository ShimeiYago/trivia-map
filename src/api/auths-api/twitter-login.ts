import { API_TIMEOUT, BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { mockLoginResponse } from 'api/mock/auths-response/login';
import { TwitterAccessTokenResponse } from './twitter-access-token';
import { LoginResponse, ValidationError } from './login';

export async function twitterLogin(param: TwitterAccessTokenResponse): Promise<LoginResponse> {
  const axiosInstance = getAxiosInstance({ timeout: API_TIMEOUT.long }, mockLoginResponse);

  const requestData: TwitterLoginRequest = {
    access_token: param.accessToken,
    token_secret: param.accessTokeSecret,
  };

  try {
    const res: AxiosResponse<LoginResponse> = await axiosInstance.post(
      `${BASE_URL}/auths/twitter/login`,
      requestData,
    );
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type TwitterLoginRequest = {
  access_token: string;
  token_secret: string;
};
