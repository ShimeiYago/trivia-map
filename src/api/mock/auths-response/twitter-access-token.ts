import { TwitterAccessTokenResponse } from './../../auths-api/twitter-access-token';

export const mockTwitterAccessTokenResponse: TwitterAccessTokenResponse = {
  accessToken: 'xxx',
  accessTokenSecret: 'xxx',
};
