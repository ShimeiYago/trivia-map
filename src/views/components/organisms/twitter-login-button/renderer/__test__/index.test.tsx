import { shallow, ShallowWrapper } from 'enzyme';
import { Renderer, Props, State } from '..';
import * as TwitterRequestTokenModule from 'api/auths-api/twitter-request-token';
import { mockTwitterRequestTokenResponse } from 'api/mock/auths-response/twitter-request-token';
import * as TwitterLoginnModule from 'api/auths-api/twitter-login';
import { mockLoginResponse } from 'api/mock/auths-response/login';
import { mockTwitterAccessTokenResponse } from 'api/mock/auths-response/twitter-access-token';
import { TwitterAccessTokenResponse } from 'api/auths-api/twitter-access-token';

let wrapper: ShallowWrapper<Props, State, Renderer>;
let twitterRequestTokenSpy: jest.SpyInstance;
let twitterLoginSpy: jest.SpyInstance;

const basicProps: Props = {
  throwError: jest.fn(),
  loginSuccess: jest.fn(),
  setAccessTokenExpiration: jest.fn(),
  setRefreshTokenExpiration: jest.fn(),
};

describe('Shallow Snapshot Tests', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('basic', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('handleClick', () => {
  beforeEach(() => {
    twitterRequestTokenSpy = jest.spyOn(TwitterRequestTokenModule, 'twitterRequestToken');
    wrapper = shallow(<Renderer {...basicProps} />);
    jest.spyOn(window, 'open').mockResolvedValue({} as never);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should open window', async () => {
    twitterRequestTokenSpy.mockResolvedValue(mockTwitterRequestTokenResponse);

    const instance = wrapper.instance();
    await instance['handleClick']();

    expect(instance.props.throwError).not.toBeCalled();
  });

  it('should throw error if api is failed', async () => {
    twitterRequestTokenSpy.mockRejectedValue({});

    const instance = wrapper.instance();

    await instance['handleClick']();

    expect(instance.props.throwError).toBeCalled();
  });
});

describe('componentWillUnmount', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should set null as onmessage', async () => {
    const instance = wrapper.instance();

    instance['componentWillUnmount']();

    expect(window.onmessage).toBeNull();
  });
});

describe('handleMessage', () => {
  const event = {
    data: mockTwitterAccessTokenResponse,
    origin: 'http://localhost',
  } as MessageEvent<TwitterAccessTokenResponse>;

  beforeEach(() => {
    twitterLoginSpy = jest.spyOn(TwitterLoginnModule, 'twitterLogin');
    wrapper = shallow(<Renderer {...basicProps} />);
    wrapper.setState({
      loading: true,
    });

    jest.spyOn(window, 'open').mockResolvedValue({} as never);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should set loading state false', async () => {
    twitterLoginSpy.mockResolvedValue(mockLoginResponse);
    const instance = wrapper.instance();

    await instance['handleMessage'](event);

    expect(instance.state.loading).toBeFalsy();
  });

  it('should throw error if api is failed', async () => {
    twitterLoginSpy.mockRejectedValue({});

    const instance = wrapper.instance();

    await instance['handleMessage'](event);

    expect(instance.props.throwError).toBeCalled();
  });

  it('should return without any process', async () => {
    const event = {
      data: mockTwitterAccessTokenResponse,
      origin: 'xxx',
    } as MessageEvent<TwitterAccessTokenResponse>;

    const instance = wrapper.instance();

    await instance['handleMessage'](event);

    expect(instance.state.loading).toBeTruthy();
  });
});
