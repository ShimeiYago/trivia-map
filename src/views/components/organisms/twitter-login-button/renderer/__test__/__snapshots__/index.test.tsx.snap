// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Snapshot Tests basic 1`] = `
<ForwardRef(LoadingButton)
  loading={false}
  onClick={[Function]}
  startIcon={<Memo />}
  variant="outlined"
>
  Twitterでログイン
</ForwardRef(LoadingButton)>
`;
